workflow-manager/limiter/limiter.go

```diff
@@ -0,0 +1,69 @@
+package limiter
+
+import (
+	"fmt"
+	"sync/atomic"
+)
+
+// Limiter is similar to sync.Pool, however, it limtis how many shared
+// objects (in this case workers) there can be
+type Limiter struct {
+	workers    chan int32
+	inProgress int32
+}
+
+// Ticket is the structure that allows you to schedule jobs
+type Ticket struct {
+	limiter *Limiter
+	worker  int32
+	dirty   int32
+}
+
+// New creates a new instance of Limiter that limits how many functions can get executed without the other ones completing first
+func New(maxWorkers int32) *Limiter {
+	l := &Limiter{make(chan int32, maxWorkers), 0}
+
+	for i := int32(0); i < maxWorkers; i++ {
+		l.workers <- i
+	}
+
+	return l
+}
+
+// RequestTicket requests a ticket for a worker
+// this call will block if there are no tickets available for work
+func (l *Limiter) RequestTicket() *Ticket {
+	worker := <-l.workers
+	return &Ticket{
+		limiter: l,
+		worker:  worker,
+		dirty:   int32(0),
+	}
+}
+
+// Execute Executes a function with the work ticket
+func (t *Ticket) Execute(job func()) error {
```

Comment:
Am I ever going to call `RequestTicket()` without immediately calling `Execute()`? I think you could simplify the interface by having a single method `limiter.Execute(job func()) (*Ticket, error)` that blocks on an available worker (by reading from `l.workers`) and then executes the provided `job`. Then you don't need the atomic test on `ticket.Dirty` because there is no way to call `Execute()` twice on a single `Ticket`.

Reply:
I see your point. I would do `func (l *Limiter) Execute(job func(*Ticket)) error`, with the contract that `job` (or some other goroutine started by `job`) is expected to call `ticket.Done()`.